app/src/main/java/com/sopt/androidstudy/presentation/save/viewmodels/FriendViewModel.kt

```diff
@@ -0,0 +1,88 @@
+package com.sopt.androidstudy.presentation.save.viewmodels
+
+import android.util.Patterns
+import android.widget.Toast
+import androidx.lifecycle.*
+import com.google.android.material.snackbar.Snackbar
+import com.sopt.androidstudy.data.model.db.Friend
+import com.sopt.androidstudy.data.model.db.FriendRepository
+import com.sopt.androidstudy.presentation.util.Event
+import kotlinx.coroutines.launch
+
+class FriendViewModel(private val repository: FriendRepository) : ViewModel() {
+
+    val friends = repository.friends
+    val position = MutableLiveData<Int?>()
+    val inputName = MutableLiveData<String?>()
+    val inputEmail = MutableLiveData<String?>()
+    val saveOrUpdateButtonText = MutableLiveData<String>()
+    val clearAllOrDeleteButtonText = MutableLiveData<String>()
+
+    val isValid = MutableLiveData<Boolean>()
+    private val _showToast = MutableLiveData<Event<Boolean>>()
+    val showToast: LiveData<Event<Boolean>> = _showToast
+
+    init {
+        initSetting()
+    }
+
+    fun saveOrUpdate() {
+        if (!inputName.value.isNullOrBlank() && !inputEmail.value.isNullOrBlank()) {
+            val name = inputName.value!!
+            val email = inputEmail.value!!
+            if (position.value == null) {
+                insert(Friend(0, name, email))
+            } else {
+                update(Friend(friends.value!!.get(position.value!!).id, name, email))
+                initSetting()
+            }
+
+        }
+    }
+
+    fun onStartEvent() {
+        _showToast.value = Event(true)
+    }
+
+    fun clearAllOrDelete() {
+        if (position.value == null) clearAll() else {
+            delete(friends.value!!.get(position.value!!))
+            initSetting()
+        }
+    }
+
+    fun insert(friend: Friend) {
+        viewModelScope.launch {
+            isValid.value = repository.insert(friend)
+            onStartEvent()
+            initSetting()
+        }
+
+    }
```

Comment:
```suggestion
    fun insert(friend: Friend) {
        viewModelScope.launch(Dispatchers.IO) {
            isValid.value = repository.insert(friend)
            onStartEvent()
        }
    }
```

`saveOrUpdate()`에서 update와 insert 실행 후 `initSetting()`을 호출하면 insert에서 `initSetting()`을 삭제할 수 있겠네요~!

그리고 Dispatchers를 지정해주지 않으면 default Dispatchers인 **메인스레드**에서 실행됩니다! 
네트워크, 디스크, DB 작업에 적합한 `Dispatchers.IO`로 지정하는 것이 좋겠습니다!!

Reply:
좋은 생각입니다! 코루틴을 신경 못쓰고 있었네요!